cmd/booster-bitswap/blockfilter/blockfilter.go

```diff
@@ -0,0 +1,244 @@
+	"github.com/multiformats/go-multibase"
+)
+
+var log = logging.Logger("booster-bitswap")
+
+// BadBitsDenyList is the URL for well known bad bits list
+const BadBitsDenyList string = "https://badbits.dwebops.pub/denylist.json"
+
+// UpdateInterval is the default interval at which the public list is refected and updated
+const UpdateInterval = 5 * time.Minute
+
+// DenyListFetcher is a function that fetches a deny list in the json style of the BadBits list
+// The first return value indicates whether any update has occurred since the last fetch time
+// The second return is a stream of data if an update has occurred
+// The third is any error
+type DenyListFetcher func(lastFetchTime time.Time) (bool, io.ReadCloser, error)
+
+const expectedListGrowth = 128
+
+// FetchBadBitsList is the default function used to get the BadBits list
+func FetchBadBitsList(ifModifiedSince time.Time) (bool, io.ReadCloser, error) {
+	req, err := http.NewRequest("GET", BadBitsDenyList, nil)
+	if err != nil {
+		return false, nil, err
+	}
+	// set the modification sync header, assuming we are not given time zero
+	if !ifModifiedSince.IsZero() {
+		req.Header.Set("If-Modified-Since", ifModifiedSince.Format(http.TimeFormat))
+	}
+	response, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return false, nil, err
+	}
+	if response.StatusCode == http.StatusNotModified {
+		return false, nil, nil
+	}
+	if response.StatusCode < 200 && response.StatusCode > 299 {
+		return false, nil, errors.New("http status code was not success")
+	}
+	return true, response.Body, nil
+}
+
+// BlockFilter manages updating a deny list and checking for CID inclusion in that list
+type BlockFilter struct {
+	cacheFile        string
+	lastUpdated      time.Time
+	denyListFetcher  DenyListFetcher
+	filteredHashesLk sync.RWMutex
+	filteredHashes   map[string]struct{}
+	ctx              context.Context
+	cancel           context.CancelFunc
+	clock            clock.Clock
+	onTimerSet       func()
+}
+
+func newBlockFilter(cfgDir string, denyListFetcher DenyListFetcher, clock clock.Clock, onTimerSet func()) *BlockFilter {
+	return &BlockFilter{
+		cacheFile:       filepath.Join(cfgDir, "denylist.json"),
+		denyListFetcher: denyListFetcher,
+		filteredHashes:  make(map[string]struct{}),
+		clock:           clock,
+		onTimerSet:      onTimerSet,
+	}
+}
+
+// NewBlockFilter returns a block filter
+func NewBlockFilter(cfgDir string) *BlockFilter {
+	return newBlockFilter(cfgDir, FetchBadBitsList, clock.New(), nil)
+}
+
+// Start initializes asynchronous updates to the deny list filter
+// It blocks to confirm at least one synchronous update of the denylist
+func (bf *BlockFilter) Start(ctx context.Context) error {
+	bf.ctx, bf.cancel = context.WithCancel(ctx)
+	// open the cache file if it eixsts
+	cache, err := os.Open(bf.cacheFile)
+	var cachedCopy bool
+	// if the file does not exist, synchronously fetch the list
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return err
```

Comment:
It could be helpful to include the context of the error here:
```suggestion
			return fmt.Errorf("fetching badbits list: %w", err)
```